tests/msc3030_test.go

```diff
@@ -0,0 +1,270 @@
+		})
+
+		t.Run("should find nothing before the earliest timestmap", func(t *testing.T) {
+			t.Parallel()
+			timeBeforeRoomCreation := time.Now()
+			roomID, _, _ := createTestRoom(t, alice)
+			mustCheckEventisReturnedForTime(t, alice, roomID, timeBeforeRoomCreation, "b", "")
+		})
+
+		t.Run("should find nothing after the latest timestmap", func(t *testing.T) {
+			t.Parallel()
+			roomID, _, eventB := createTestRoom(t, alice)
+			mustCheckEventisReturnedForTime(t, alice, roomID, eventB.AfterTimestamp, "f", "")
+		})
+
+		// Just a sanity check that we're not leaking anything from the `/timestamp_to_event` endpoint
+		t.Run("should not be able to query a private room you are not a member of", func(t *testing.T) {
+			t.Parallel()
+			timeBeforeRoomCreation := time.Now()
+
+			// Alice will create the private room
+			roomID := alice.CreateRoom(t, map[string]interface{}{
+				"preset": "private_chat",
+			})
+
+			// We will use Bob to query the room they're not a member of
+			nonMemberUser := deployment.Client(t, "hs1", "@bob:hs1")
+
+			// Make the `/timestamp_to_event` request from Bob's perspective (non room member)
+			timestamp := makeTimestampFromTime(timeBeforeRoomCreation)
+			timestampString := strconv.FormatInt(timestamp, 10)
+			timestampToEventRes := nonMemberUser.DoFunc(t, "GET", []string{"_matrix", "client", "unstable", "org.matrix.msc3030", "rooms", roomID, "timestamp_to_event"}, client.WithContentType("application/json"), client.WithQueries(url.Values{
+				"ts":  []string{timestampString},
+				"dir": []string{"f"},
+			}))
+
+			// A random user is not allowed to query for events in a private room
+			// they're not a member of (forbidden).
+			if timestampToEventRes.StatusCode != 403 {
+				t.Fatalf("/timestamp_to_event returned %d HTTP status code but expected %d", timestampToEventRes.StatusCode, 403)
+			}
+		})
+
+		t.Run("federation", func(t *testing.T) {
+			t.Run("looking forwards, should be able to find event that was sent before we joined", func(t *testing.T) {
+				t.Parallel()
+				roomID, eventA, _ := createTestRoom(t, alice)
+				remoteCharlie.JoinRoom(t, roomID, []string{"hs1"})
+				mustCheckEventisReturnedForTime(t, remoteCharlie, roomID, eventA.BeforeTimestamp, "f", eventA.EventID)
+			})
+
+			t.Run("looking backwards, should be able to find event that was sent before we joined", func(t *testing.T) {
+				t.Parallel()
+				roomID, _, eventB := createTestRoom(t, alice)
+				remoteCharlie.JoinRoom(t, roomID, []string{"hs1"})
+				mustCheckEventisReturnedForTime(t, remoteCharlie, roomID, eventB.AfterTimestamp, "b", eventB.EventID)
+			})
+		})
+	})
+}
+
+type eventTime struct {
+	EventID         string
+	BeforeTimestamp time.Time
+	AfterTimestamp  time.Time
+}
+
+func createTestRoom(t *testing.T, c *client.CSAPI) (roomID string, eventA, eventB *eventTime) {
+	t.Helper()
+
+	roomID = c.CreateRoom(t, map[string]interface{}{})
```

Comment:
Always set a `preset` - Synapse/Dendrite disagree on what the preset is. Synapse makes public rooms by default, Dendrite makes private (following the spec).